src/main/kotlin/gg/ingot/iron/transformer/ValueTransformer.kt

```diff
@@ -1,155 +1,182 @@
 
+import gg.ingot.iron.Iron
+import gg.ingot.iron.annotations.Model
 import gg.ingot.iron.representation.EntityField
-import gg.ingot.iron.serialization.*
-import gg.ingot.iron.strategies.NamingStrategy
+import gg.ingot.iron.serialization.ColumnDeserializer
+import gg.ingot.iron.serialization.ColumnSerializer
+import gg.ingot.iron.transformer.adapter.*
 import java.sql.ResultSet
+import java.sql.SQLException
 import java.util.*
-import kotlin.reflect.KClass
 
 /**
- * Transforms a value from a [ResultSet] into its corresponding type.
- * @author DebitCardz
- * @since 1.2
+ * Handles the conversion of any value to and from [ResultSet]s / prepared statements.
+ * @author santio
+ * @since 2.0
  */
-internal class ValueTransformer(
-    private val serializationAdapter: SerializationAdapter?
-) {
-    private val arrayTransformations: Map<KClass<*>, (Array<*>) -> Collection<*>> = mapOf(
-        List::class to { it.toList() },
-        Set::class to { it.toSet() }
-    )
+internal class ValueTransformer(private val iron: Iron) {
 
     /**
-     * Retrieve the value from the result set for the given field.
-     * Will automatically convert an [Array] into a given [Collection] type if the field is said [Collection].
+     * Takes a value from a [ResultSet] and converts it's appropriate java type.
      * @param resultSet The result set to retrieve the value from.
      * @param field The field to retrieve the value for.
      * @return The value from the result set.
      */
-    fun convert(
-        resultSet: ResultSet,
-        field: EntityField,
-        namingStrategy: NamingStrategy
-    ): Any? {
-        return if(field.deserializer != null) toCustomDeserializedObj(resultSet, field, namingStrategy)
-        else if(field.isJson) toJsonObject(resultSet, field, namingStrategy)
-        else if(field.isArray) toArray(resultSet, field, namingStrategy)
-        else if(field.isCollection) toCollection(resultSet, field, namingStrategy)
-        else toObject(resultSet, field, namingStrategy)
-    }
-
-    /**
-     * Retrieve the value as an [Array] from the result set.
-     * @param resultSet The result set to retrieve the value from.
-     * @param field The field to retrieve the value for.
-     * @param namingStrategy The naming strategy to use for the field.
-     * @return The array from the result set.
-     */
-    @Suppress("UNCHECKED_CAST")
-    private fun toArray(resultSet: ResultSet, field: EntityField, namingStrategy: NamingStrategy): Any? {
-        val arr = resultSet.getArray(field.convertedName(namingStrategy))
-            ?.array
-
-        if(
-            arr is Array<*>
-            && field.isEnum
-            && field.isArray
-        ) {
-            // map to enum values
-            return arr.map {
-                java.lang.Enum.valueOf(field.field.type.componentType as Class<out Enum<*>>, it as String)
-            }.toTypedArray()
+    fun fromResultSet(resultSet: ResultSet, field: EntityField): Any? {
```

Comment:
why can't we just use the `error` function?

Reply:
Stale comment